cli/docs/jqawk.1
.\" Automatically generated by Pandoc 3.9.0.2
.\"
.TH "JQAWK" "1" "" ""
.SH NAME
jqawk \- awk + JSON
.SH SYNOPSIS
\f[CR]jqawk [flags] <program> <file>...\f[R]
.SH OPTIONS
.TP
\f[B]\f[CB]\-e\f[B]\f[R] \f[CR]EXPRESSION\f[R]
Evaluate \f[CR]EXPRESSION\f[R] and print the result.
\f[B]\f[CB]$\f[B]\f[R] is set to the root value of the input file.
.TP
\f[B]\f[CB]\-f\f[B]\f[R] \f[CR]FILE\f[R]
Run the jqawk program in \f[CR]FILE\f[R]
.TP
\f[B]\f[CB]\-i\f[B]\f[R]
Start the interactive REPL
.TP
\f[B]\f[CB]\-o\f[B]\f[R] \f[CR]FILE\f[R]
Write JSON to \f[CR]FILE\f[R].
Use \f[CR]\-\f[R] to write to stdout
.TP
\f[B]\f[CB]\-c\f[B]\f[R]
Output compact JSON when the \f[B]\f[CB]\-o\f[B]\f[R] flag is given
.TP
\f[B]\f[CB]\-r\f[B]\f[R] \f[CR]SELECTOR\f[R]
Select the root value.
\f[CR]SELECTOR\f[R] is any valid expression
.TP
\f[B]\f[CB]\-\-version\f[B]\f[R]
Print version information
.TP
\f[B]\f[CB]\-\-profile\f[B]\f[R]
Record a cpu profile in \f[CR]jqawk.prof\f[R]
.TP
\f[B]\f[CB]\-\-dbg\-ast\f[B]\f[R]
Print the AST of the program and exit
.TP
\f[B]\f[CB]\-\-dbg\-lex\f[B]\f[R]
Print the parsed tokens of the program and exit
.SH DESCRIPTION
\f[B]jqawk\f[R] is an awk\-inspired programming language for wrangling
JSON.
.SS Program structure
A jqawk program is a series of rules, of the form
.IP
.EX
pattern { body }
.EE
.PP
\f[CR]pattern\f[R] can be any expression, and \f[CR]body\f[R] is a
sequence of statements.
If \f[CR]pattern\f[R] evaluates to a truthy value, \f[CR]body\f[R] is
evaluated.
.PP
There are also some special patterns:
.PP
\f[B]\f[CB]BEGIN\f[B]\f[R] and \f[B]\f[CB]END\f[B]\f[R] run before and
after processing all files
.PD 0
.P
.PD
\f[B]\f[CB]BEGINFILE\f[B]\f[R] and \f[B]\f[CB]ENDFILE\f[B]\f[R] run
before after processing each file
.PP
All other rules are run for each value in the file.
.SS Variables
Some variables are pre\-defined in the body of a rule.
If the input is an array:
.PP
\f[B]\f[CB]$\f[B]\f[R] is the current element
\f[B]\f[CB]$index\f[B]\f[R] is the array index of that element
.PP
If the input is anything else:
.PP
\f[B]\f[CB]$\f[B]\f[R] is the root value
.SS Statements
The following statements are valid:
.TP
\f[CR]print <expression>, <expression>, ...\f[R]
print \f[CR]<expression>\f[R]s, separated by spaces
.TP
\f[CR]return <expression>\f[R]
return from the current function, optionally returning the value of
\f[CR]<expression>\f[R]
.TP
\f[CR]break\f[R]
exit the enclosing for or while loop
.TP
\f[CR]continue\f[R]
immediately begin the next iteration of the enclosing loop
.TP
\f[CR]next\f[R]
immediately exit the rule and process no further rules for the current
element
.TP
\f[CR]exit <expression>\f[R]
immediately exit the program.
the exit code is set to \f[CR]<expression>\f[R], or 0 if no expression
is present
.TP
\f[CR]if (<expression>) <body> else <elsebody>\f[R]
execute \f[CR]<body>\f[R] if \f[CR]<expression>\f[R] is truthy.
If the value is falsy and the else is present, execute
\f[CR]<elsebody>\f[R]
.TP
\f[CR]while (<expression>) <body>\f[R]
execute the body until \f[CR]<expression>\f[R] is falsy
.TP
\f[CR]for (<preexpression>; <checkexpression>; <postexpression>) <body>\f[R]
execute \f[CR]<preexpression>\f[R], then execute
\f[CR]{ <body>; <postexpression> }\f[R] until
\f[CR]<checkexpression>\f[R] is falsy
.RS
.sp
all three expressions are optional.
if \f[CR]<checkexpression>\f[R] is missing, it is taken to always be
true.
\f[CR]for (;;)\f[R] is an infinite loop
.RE
.TP
\f[CR]for (<identifier> in <expression>) <body>\f[R]
execute \f[CR]<body>\f[R] for each element in \f[CR]<expression>\f[R]
with \f[CR]<identifier>\f[R] set to its value.
.RS
.sp
if \f[CR]<expression>\f[R] is an array, \f[CR]<identifier>\f[R] is each
element.
.PD 0
.P
.PD
if it\(cqs an object, \f[CR]<identifier>\f[R] is each key.
.PD 0
.P
.PD
if it\(cqs a string, \f[CR]<identifier>\f[R] is each character.
.PD 0
.P
.PD
.RE
.TP
\f[CR]for (<identifier>, <indexidentifier> in <expression>) <body>\f[R]
same as above, with \f[CR]<indexindeitifer>\f[R] set to the index of the
element for arrays or the value for objects
.SS Built\-in functions
.TP
\f[CR]printf(format_string, args...)\f[R]
format and print a string according to the format specifier.
the following format codes are supported:
.RS
.IP
.EX
  %% \- a literal %
  %c \- the character represented by the given number
  %d \- integer
  %i \- integer
  %o \- octal integer
  %x \- hexidecimal integer
  %f \- floating\-point number
  %s \- string
  %v \- the value in its default format
.EE
.RE
.TP
\f[CR]json(arg)\f[R]
return arg converted to an indented JSON string
.TP
\f[CR]parseJson(string)\f[R]
return the string parsed as a JSON value
.TP
\f[CR]num(arg)\f[R]
convert arg to a number (64\-bit float).
returns null if the conversion fails
.SS Methods
.SS String
.TP
\f[CR]string.length()\f[R]
return the length
.TP
\f[CR]string.upper()\f[R]
return an uppercase copy of the string
.TP
\f[CR]string.lower()\f[R]
return a lowercase copy of the string
.TP
\f[CR]string.split(separator)\f[R]
split the string into substrings on a separator.
the separator may be a string or a regex literal
.TP
\f[CR]strings.trim()\f[R]
return the string with leading and trailing whitespace removed
.SS Number
.TP
\f[CR]number.floor()\f[R]
floor the given number
.TP
\f[CR]number.ceil()\f[R]
ceil or round the given number
.TP
\f[CR]number.round()\f[R]
round the given number.
halves are rounded away from zero
.TP
\f[CR]number.abs()\f[R]
return tbe absolute value of the number
.TP
\f[CR]number.mod(n)\f[R]
return the number wrapped into the range [0, n].
unlike the modulo operator, the result is never negative
.TP
\f[CR]number.sqrt(n)\f[R]
return the square root of x
.TP
\f[CR]number.format(thousandsSeparator, decimalSeparator)\f[R]
format a number with thousands and decimal separators.
if no separators are given \f[CR],\f[R] is used for thousands and
\f[CR].\f[R] is used for decimals
.SS Object
.TP
\f[CR]object.length()\f[R]
return the number of keys
.TP
\f[CR]object.pluck(k1, k2, ...)\f[R]
return a shallow copy of the object containing only the given keys
the keys can be passed as separate arguments, or as an array
.TP
\f[CR]object.keys()\f[R]
return the keys of an object as a list
.TP
\f[CR]object.values()\f[R]
return the values of an object as a list
.TP
\f[CR]object.pairs()\f[R]
return a list of [key, value] for each entry in the object
.SS Array
.TP
\f[CR]array.length()\f[R]
return the length
.TP
\f[CR]array.push(value)\f[R]
push value to the end of the array
.TP
\f[CR]array.pop()\f[R]
remove and return the last value in the array
.TP
\f[CR]array.popfirst()\f[R]
remove and return the first value in the array
.TP
\f[CR]array.contains(value)\f[R]
return true if value is in the array
.TP
\f[CR]array.sort(fn)\f[R]
return a sorted copy of the array.
.RS
.sp
with no fn given the array is sorted by string value, or numerically if
all the values are nunbers.
.sp
with fn(a, b) provided, the array is sorted according to fn\(cqs return
value:
.PD 0
.P
.PD
1 means a < b
.PD 0
.P
.PD
0 means a == b
.PD 0
.P
.PD
1 means a > b
.RE
.TP
\f[CR]array.sortKey(string|fn)\f[R]
return a sorted copy of the array sorted by a key taken from each
element.
.RS
.sp
with a string argument, the key is \f[CR]element[string]\f[R] with a
function argument, the key is the result of \f[CR]fn(element)\f[R]
.RE
.TP
\f[CR]array.reverse()\f[R]
return a reversed copy of the array
.TP
\f[CR]array.map(fn)\f[R]
return a copy of the array where each element is the result of
\f[CR]fn(element)\f[R]
.TP
\f[CR]array.filter(fn)\f[R]
return a new array that only includes elements where
\f[CR]fn(element)\f[R] is truthy
.SH SEE ALSO
Project homepage \c
.UR http://github.com/alligator/jqawk
.UE \c


jqawk.js
import { parse } from "https://deno.land/std@0.106.0/flags/mod.ts";

class Lexer {
  constructor(src) {
    this.src = src;
    this.pos = 0;
    this.line = 1;
  }

  advance() {
    if (this.pos < this.src.length) {
      this.pos++;
    }
    return this.src[this.pos - 1];
  }
  peek() {
    return this.src[this.pos];
  }
  atEOF() {
    return this.pos >= this.src.length;
  }

  simpleToken(type) {
    return { type, line: this.line };
  }
  stringToken(type, str) {
    return { type, str, line: this.line };
  }

  isAlpha(c) {
    return /^[a-zA-Z]+/.test(c);
  }
  identifier() {
    while (!this.atEOF() && this.isAlpha(this.peek())) {
      this.advance();
    }

    const word = this.src.substring(this.start, this.pos);
    switch (word) {
      case 'print':
        return this.simpleToken('print');
      case 'BEGIN':
        return this.simpleToken('begin');
      case 'END':
        return this.simpleToken('end');
      default:
        return this.stringToken('identifier', word);
    }
  }

  isNumeric(c) {
    return /^[0-9]+/.test(c);
  }
  number() {
    while (!this.atEOF() && this.isNumeric(this.peek())) {
      this.advance();
    }

    const num = this.src.substring(this.start, this.pos);
    return this.stringToken('number', num);
  }
  
  string() {
    while (this.peek() !== '"') {
      this.advance();
      if (this.atEOF()) {
        throw new Error('unexpected EOF in string');
      }
    }
    this.advance();

    const str = this.src.substring(this.start + 1, this.pos - 1);
    return this.stringToken('string', str);
  }

  skipWhitespace() {
    while (true) {
      switch (this.peek()) {
        case ' ': {
          this.advance();
          break;
        }
        case '\n': {
          this.advance();
          this.line++;
          break;
        }
        default:
          return;
      }
    }
  }

  nextToken() {
    this.skipWhitespace();
    if (this.pos >= this.src.length) {
      return this.simpleToken('eof');
    }

    this.start = this.pos;
    const c = this.advance();

    if (this.isAlpha(c)) {
      return this.identifier();
    }

    if (this.isNumeric(c)) {
      return this.number();
    }

    if (c === '"') {
      return this.string();
    }

    switch (c) {
      case '$': return this.simpleToken('dollar');
      case '+': return this.simpleToken('plus');
      case '>': return this.simpleToken('greater');
      case '{': return this.simpleToken('lcurly');
      case '}': return this.simpleToken('rcurly');
      case '[': return this.simpleToken('lsquare');
      case ']': return this.simpleToken('rsquare');
      case '.': return this.simpleToken('dot');
      case ',': return this.simpleToken('comma');
      case ';': return this.simpleToken('semicolon');
      case '=': {
        if (this.peek() === '=') {
          this.advance();
          return this.simpleToken('equalequal');
        }
        return this.simpleToken('equal');
      }
    }

    throw new Error(`unexpected character ${JSON.stringify(c)}`);
  }
}

// this "parser" does the bare minimum. it parses a list of these:
//   <stream of tokens> { <stream of tokens }
//
// the actual parser for the language is inside the evaluator
class Parser {
  constructor(src) {
    this.src = src;
    this.lexer = new Lexer(src);
  }

  advance() {
    this.current = this.lexer.nextToken();
  }

  consume(type) {
    if (this.current.type !== type) {
      throw new Error(`found ${this.current.type} but expected ${type}`);
    }
    this.advance();
  }

  parseProgram() {
    this.advance();

    const rules = {
      begin: [],
      main: [],
      end: [],
    };

    while (this.current.type !== 'eof') {
      const pattern = [];
      while (this.current.type !== 'lcurly') {
        pattern.push(this.current);
        this.advance();
      }

      this.consume('lcurly');

      const body = [];
      while (this.current.type !== 'rcurly') {
        body.push(this.current);
        this.advance();
      }

      this.consume('rcurly');

      if (pattern.length && pattern[0].type === 'begin') {
        rules.begin.push({ pattern, body });
      } else if (pattern.length && pattern[0].type === 'end') {
        rules.end.push({ pattern, body });
      } else {
        rules.main.push({ pattern, body });
      }
    }

    return rules;
  }
}

function val(value) {
  if (value === null) {
    return { type: null };
  }
  switch (typeof value) {
    case 'number':
      return { type: 'number', value };
    case 'string':
      return { type: 'string', value };
    case 'boolean':
      return { type: 'bool', value };
    case 'object': {
      if (Array.isArray(value)) {
        return { type: 'array', value };
      }
      return { type: 'object', value };
    }
  }

  throw new Error(`cannot create a value from ${value}`);
}

class Evaluator {
  constructor(prog) {
    this.prog = prog;
    this.pos = 0;
    this.environment = {};
    this.fields = {};
  }

  // hand rolled pratt parser
  // this should probably be in it's own class...
  evaluate(environment, fields, tokens, startType) {
    const precedence = {
      none: 0,
      assign: 10,
      seq: 20,
      comp: 30,
      add: 40,
      mul: 50,
      fn: 60,
    };

    let current = tokens[0];
    let pos = 1;
    let prev;

    // utils
    const advance = () => {
      prev = current;
      current = tokens[pos++];
      if (!current) {
        current = { type: 'eof', line: prev.line };
      }
    };

    const consume = (type) => {
      if (current.type !== type) {
        fatal(`expected ${type} but got ${current.type}`);
      }
      advance();
    };

    const fatal = (msg) => {
      throw new Error(`error on line ${current.line}: ${msg}`);
    };

    // grammar
    const printStatement = () => {
      consume('print');
      const args = [];
      while (current.type !== 'semicolon') {
        args.push(expression());
        if (current.type === 'comma') {
          consume('comma');
        } else {
          break;
        }
      }

      console.log(args.map((arg) => {
        switch (arg.type) {
          case 'array':
            return Deno.inspect(arg.value);
          case 'object':
            return Deno.inspect(arg.value);
          case null:
            return '<null>';
          default:
            return arg.value;
        }
      }).join(' '));
    };

    const statement = () => {
      let result;
      switch (current.type) {
        case 'print': {
          result = printStatement();
          break;
        }
        default: {
          result = expression();
          break;
        }
      }

      consume('semicolon');
      return result;
    };

    const number = () => {
      return val(parseInt(prev.str, 10));
    };
    const string = () => {
      return val(prev.str);
    };

    const member = (left) => {
      consume('dot');
      consume('identifier');
      if (left.type !== 'object') {
        fatal('cannot access member of non object');
      }
      return val(left.value[prev.str]);
    };

    const coerceTypes = (left, right) => {
      if (left.type === right.type) {
        return;
      }

      if (left.type === null) {
        // create a default
        left.type = right.type;
        switch (right.type) {
          case 'number':
            left.value = 0;
            break;
          case 'string':
            left.value = '';
            break;
          case 'array':
            left.value = [];
            break;
          case 'object':
            left.value = {};
            break;
        }
        return;
      }
    }

    const binary = (left) => {
      const token = current;
      advance();

      const right = expression(getRule(token.type).prec);

      switch (token.type) {
        case 'greater':
          return val(left.value > right.value);
        case 'equalequal':
          return val(left.value === right.value);
        case 'plus':
          coerceTypes(left, right);
          return val(left.value + right.value);
      }

      fatal(`unknown operator ${token.type}`);
    };

    const identifier = () => {
      const key = prev.str;
      if (!(key in environment)) {
        environment[key] = val(null);
      }
      return environment[key];
    };

    const field = () => {
      let key = 'root';
      if (current.type === 'identifier') {
        consume('identifier');
        key = prev.str;
      }

      if (!(key in fields)) {
        fatal(`unknown field ${key}`);
      }

      return fields[key];
    };

    const subscript = (left) => {
      consume('lsquare');
      const key = expression();
      consume('rsquare');

      if (left.type !== 'array') {
        fatal('cannot subscript non array value');
      }

      return val(left.value[key.value]);
    };

    const assign = (left) => {
      consume('equal');
      const value = expression();
      left.type = value.type;
      left.value = value.value;
      return left;
    };

    const getRule = (type) => {
      const r = {
        number: {
          prec: precedence.none,
          prefix: number,
        },
        string: {
          prec: precedence.none,
          prefix: string,
        },
        dot: {
          prec: precedence.fn,
          infix: member,
        },
        greater: {
          prec: precedence.comp,
          infix: binary,
        },
        dollar: {
          prec: precedence.none,
          prefix: field,
        },
        equalequal: {
          prec: precedence.comp,
          infix: binary,
        },
        plus: {
          prec: precedence.add,
          infix: binary,
        },
        lsquare: {
          prec: precedence.fn,
          infix: subscript,
        },
        identifier: {
          prec: precedence.none,
          prefix: identifier,
        },
        equal: {
          prec: precedence.assign,
          infix: assign,
        },
      }[type] || { prec: precedence.none };
      return r;
    };

    const expression = (prec = precedence.assign) => {
      const token = current;
      const rule = getRule(token.type);
      if (!rule.prefix) {
        fatal(`unexpected prefix ${token.type}`);
      }

      advance();
      let left = rule.prefix();

      while (prec <= getRule(current.type).prec) {
        const infixRule = getRule(current.type);
        if (!infixRule.infix) {
          fatal(`unexpected infix ${current.type}`);
        }
        left = infixRule.infix(left);
      }

      return left;
    };

    let result;
    if (startType === 'expression') {
      result = expression(precedence.assign);
    } else {
      while (current.type !== 'eof') {
        statement();
      }
    }

    if (pos < tokens.length) {
      fatal(`dangling tokens ${JSON.stringify(tokens.slice(pos, pos + 5))}`);
    }
    return result;
  }

  forEachRecord(json, cb) {
    if (Array.isArray(json)) {
      while (this.pos < json.length) {
        this.fields.key = val(this.pos);
        cb(json[this.pos++]);
      }
      return;
    }

    if (typeof json === 'object') {
      Object.keys(json).forEach((key) => {
        this.fields.key = val(key);
        cb(json[key]);
      });
      return;
    }

    throw new Error('expected top level JSON to be an array or object');
  }

  getRoot(selector, json) {
    if (!selector) {
      return val(json);
    }
    const lexer = new Lexer(selector);
    const tokens = [];
    while (true) {
      const token = lexer.nextToken();
      if (token.type === 'eof') {
        break;
      }
      tokens.push(token);
    }

    const root = this.evaluate(
      {},
      { root: val(json) },
      tokens,
      'expression',
    );

    return root;
  }

  async run(selector, json) {
    const root = this.getRoot(selector, json);

    this.prog.begin.forEach(({ body }) => {
      this.evaluate(this.environment, this.fields, body, 'statement');
    });

    if (root) {
      this.forEachRecord(root.value, (record) => {
        this.prog.main.forEach(({ pattern, body }) => {
          this.fields.root = val(record);
          let result;
          if (pattern.length === 0) {
            result = { value: true };
          } else {
            result = this.evaluate(
              this.environment,
              this.fields,
              pattern,
              'expression',
            );
          }
          if (result.value) {
            this.evaluate(
              this.environment,
              this.fields,
              body,
              'statement',
            );
          }
        });
      });
    }

    this.prog.end.forEach(({ body }) => {
      this.evaluate(
        this.environment,
        this.fields,
        body,
        'statement',
      );
    });
  }
}

function usage() {
  console.log('usage: jqawk -f program_file [args] file');
  console.log('       jqawk \'program\' [args] file');
  console.log('');
  console.log('arguments: -s  selector');
  console.log('           -h  display this message');
  Deno.exit(0);
}

const args = parse(Deno.args);

if (args.h || args.help || args._.length < 1) {
  usage();
}

let src;
let file;
let selector = args.s;

if (args.f) {
  src = Deno.readTextFileSync(args.f);
  file = args._[0];
} else {
  if (args._.length < 1) {
    usage()
  }
  src = args._[0];
  file = args._[1];
}

const p = new Parser(src);
const prog = p.parseProgram();

try {
  const e = new Evaluator(prog);

  if (file) {
    const json = JSON.parse(await Deno.readTextFile(file));
    await e.run(selector, json);
  } else {
    const stdin = await Deno.readAll(Deno.stdin);
    const content = new TextDecoder().decode(stdin);
    const json = JSON.parse(content);
    await e.run(selector, json);
  }

} catch (e) {
  console.error(e.message);
  Deno.exit(1);
}
